File: codemem/raw_event_flush.py
from __future__ import annotations

import datetime as dt
import logging
import os
from typing import Any

from . import plugin_ingest
from .observer import ObserverAuthError
from .store import MemoryStore

EXTRACTOR_VERSION = "raw_events_v1"
logger = logging.getLogger(__name__)
ingest = plugin_ingest.ingest


def _truncate_error_message(message: str, *, limit: int = 280) -> str:
    text = " ".join(message.split())
    if len(text) <= limit:
        return text
    return f"{text[: limit - 3].rstrip()}..."


def _provider_display_name(provider: str | None) -> str:
    normalized = (provider or "").strip().lower()
    if normalized == "openai":
        return "OpenAI"
    if normalized == "anthropic":
        return "Anthropic"
    if normalized:
        return normalized.capitalize()
    return "Observer"


def _summarize_flush_failure(exc: Exception, *, provider: str | None) -> str:
    provider_title = _provider_display_name(provider)
    raw_message = str(exc).strip().lower()
    if isinstance(exc, ObserverAuthError):
        return f"{provider_title} authentication failed. Refresh credentials and retry."
    if isinstance(exc, TimeoutError):
        return f"{provider_title} request timed out during raw-event processing."
    if raw_message == "observer failed during raw-event flush":
        return f"{provider_title} returned no usable output for raw-event processing."
    if "parse" in raw_message or "xml" in raw_message or "json" in raw_message:
        return f"{provider_title} response could not be processed."
    return f"{provider_title} processing failed during raw-event ingestion."


def _flush_failure_details(exc: Exception) -> dict[str, str | None]:
    observer = plugin_ingest.OBSERVER
    active = observer.get_status() if observer is not None else None
    provider = (
        str(active.get("provider") or "").strip() or None if isinstance(active, dict) else None
    )
    model = str(active.get("model") or "").strip() or None if isinstance(active, dict) else None
    runtime = str(active.get("runtime") or "").strip() or None if isinstance(active, dict) else None
    last_error = active.get("last_error") if isinstance(active, dict) else None
    last_error_message = None
    last_error_code = None
    if isinstance(last_error, dict):
        last_error_message = str(last_error.get("message") or "").strip() or None
        last_error_code = str(last_error.get("code") or "").strip() or None
    logger.warning(
        "raw event flush failed",
        extra={
            "observer_provider": provider or "unknown",
            "observer_model": model or "unknown",
            "observer_runtime": runtime or "unknown",
            "error_type": exc.__class__.__name__,
        },
        exc_info=exc,
    )
    return {
        "message": _truncate_error_message(
            last_error_message or _summarize_flush_failure(exc, provider=provider)
        ),
        "error_type": last_error_code or exc.__class__.__name__,
        "observer_provider": provider,
        "observer_model": model,
        "observer_runtime": runtime,
    }


def build_session_context(events: list[dict[str, Any]]) -> dict[str, Any]:
    prompt_count = sum(1 for e in events if e.get("type") == "user_prompt")
    tool_count = sum(1 for e in events if e.get("type") == "tool.execute.after")

    ts_values = []
    for e in events:
        ts = e.get("timestamp_wall_ms")
        if ts is None:
            continue
        try:
            ts_values.append(int(ts))
        except (TypeError, ValueError):
            continue
    duration_ms = 0
    if ts_values:
        duration_ms = max(0, max(ts_values) - min(ts_values))

    files_modified: set[str] = set()
    files_read: set[str] = set()
    for e in events:
        if e.get("type") != "tool.execute.after":
            continue
        tool = str(e.get("tool") or "").lower()
        args = e.get("args") or {}
        if not isinstance(args, dict):
            continue
        file_path = args.get("filePath") or args.get("path")
        if not isinstance(file_path, str) or not file_path:
            continue
        if tool in {"write", "edit"}:
            files_modified.add(file_path)
        if tool == "read":
            files_read.add(file_path)

    first_prompt = None
    for e in events:
        if e.get("type") != "user_prompt":
            continue
        text = e.get("prompt_text")
        if isinstance(text, str) and text.strip():
            first_prompt = text.strip()
            break

    return {
        "first_prompt": first_prompt,
        "prompt_count": prompt_count,
        "tool_count": tool_count,
        "duration_ms": duration_ms,
        "files_modified": sorted(files_modified),
        "files_read": sorted(files_read),
    }


def flush_raw_events(
    store: MemoryStore,
    *,
    opencode_session_id: str,
    source: str = "opencode",
    cwd: str | None,
    project: str | None,
    started_at: str | None,
    max_events: int | None = None,
) -> dict[str, int]:
    def _event_seq(value: Any, fallback: int) -> int:
        if value is None:
            return fallback
        try:
            return int(value)
        except (TypeError, ValueError):
            return fallback

    def _event_seq_opt(value: Any) -> int | None:
        if value is None:
            return None
        try:
            return int(value)
        except (TypeError, ValueError):
            return None

    source = source.strip().lower() or "opencode"
    meta = store.raw_event_session_meta(opencode_session_id, source=source)
    if cwd is None:
        cwd = meta.get("cwd") or os.getcwd()
    if project is None:
        project = meta.get("project")
    if started_at is None:
        started_at = meta.get("started_at")

    last_flushed = store.raw_event_flush_state(opencode_session_id, source=source)
    events = store.raw_events_since_by_seq(
        opencode_session_id=opencode_session_id,
        source=source,
        after_event_seq=last_flushed,
        limit=max_events,
    )
    if not events:
        return {"flushed": 0, "updated_state": 0}

    event_seqs = [_event_seq_opt(e.get("event_seq")) for e in events]
    event_seqs = [seq for seq in event_seqs if seq is not None]
    if not event_seqs:
        return {"flushed": 0, "updated_state": 0}

    start_event_seq = min(event_seqs)
    last_event_seq = max(event_seqs)
    if last_event_seq < start_event_seq:
        return {"flushed": 0, "updated_state": 0}

    batch_id, status = store.get_or_create_raw_event_flush_batch(
        opencode_session_id=opencode_session_id,
        source=source,
        start_event_seq=start_event_seq,
        end_event_seq=last_event_seq,
        extractor_version=EXTRACTOR_VERSION,
    )
    if status == "completed":
        store.update_raw_event_flush_state(opencode_session_id, last_event_seq, source=source)
        return {"flushed": 0, "updated_state": 1}

    if not store.claim_raw_event_flush_batch(batch_id):
        return {"flushed": 0, "updated_state": 0}
    session_context = build_session_context(events)
    session_context["opencode_session_id"] = opencode_session_id
    session_context["source"] = source
    session_context["stream_id"] = opencode_session_id
    session_context["start_event_seq"] = start_event_seq
    session_context["end_event_seq"] = last_event_seq
    session_context["flusher"] = "raw_events"
    session_context["extractor_version"] = EXTRACTOR_VERSION
    session_context["flush_batch"] = {
        "batch_id": batch_id,
        "start_event_seq": start_event_seq,
        "end_event_seq": last_event_seq,
    }

    payload = {
        "cwd": cwd,
        "project": project,
        "started_at": started_at or dt.datetime.now(dt.UTC).isoformat(),
        "events": events,
        "session_context": session_context,
    }
    try:
        ingest(payload)
    except Exception as exc:
        details = _flush_failure_details(exc)
        store.record_raw_event_flush_batch_failure(
            batch_id,
            message=str(details["message"] or exc.__class__.__name__),
            error_type=str(details["error_type"] or exc.__class__.__name__),
            observer_provider=details["observer_provider"],
            observer_model=details["observer_model"],
            observer_runtime=details["observer_runtime"],
        )
        raise
    store.update_raw_event_flush_batch_status(batch_id, "completed")
    store.update_raw_event_flush_state(opencode_session_id, last_event_seq, source=source)
    return {"flushed": len(events), "updated_state": 1}

File: packages/core/src/raw-event-flush.ts
/**
 * Raw event flush orchestration — bridge between the sweeper and the ingest pipeline.
 *
 * Ports the core flush logic from codemem/raw_event_flush.py.
 *
 * Reads unflushed raw events for a session, creates a batch record,
 * builds an IngestPayload, runs it through the ingest pipeline,
 * and updates flush state on success (or records failure details).
 */

import { type IngestOptions, ingest } from "./ingest-pipeline.js";
import type { IngestPayload, SessionContext } from "./ingest-types.js";
import { ObserverAuthError } from "./observer-client.js";
import type { MemoryStore } from "./store.js";

const EXTRACTOR_VERSION = "raw_events_v1";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function truncateErrorMessage(message: string, limit = 280): string {
	const text = message.replace(/\s+/g, " ").trim();
	if (text.length <= limit) return text;
	return `${text.slice(0, limit - 3).trimEnd()}...`;
}

function providerDisplayName(provider: string | null | undefined): string {
	const normalized = (provider ?? "").trim().toLowerCase();
	if (normalized === "openai") return "OpenAI";
	if (normalized === "anthropic") return "Anthropic";
	if (normalized) return normalized.charAt(0).toUpperCase() + normalized.slice(1);
	return "Observer";
}

function summarizeFlushFailure(exc: Error, provider: string | null | undefined): string {
	const providerTitle = providerDisplayName(provider);
	const rawMessage = String(exc.message ?? "")
		.trim()
		.toLowerCase();

	if (exc instanceof ObserverAuthError) {
		return `${providerTitle} authentication failed. Refresh credentials and retry.`;
	}
	if (exc.name === "TimeoutError" || rawMessage.includes("timeout")) {
		return `${providerTitle} request timed out during raw-event processing.`;
	}
	if (
		rawMessage === "observer failed during raw-event flush" ||
		rawMessage === "observer produced no storable output for raw-event flush"
	) {
		return `${providerTitle} returned no usable output for raw-event processing.`;
	}
	if (/parse|xml|json/i.test(rawMessage)) {
		return `${providerTitle} response could not be processed.`;
	}
	return `${providerTitle} processing failed during raw-event ingestion.`;
}

// ---------------------------------------------------------------------------
// Session context builder
// ---------------------------------------------------------------------------

/**
 * Build session context from raw events — extracts prompt count, tool count,
 * duration, files modified/read, and first user prompt.
 *
 * Port of build_session_context() from raw_event_flush.py.
 */
export function buildSessionContext(events: Record<string, unknown>[]): SessionContext {
	let promptCount = 0;
	let toolCount = 0;

	for (const e of events) {
		if (e.type === "user_prompt") promptCount++;
		if (e.type === "tool.execute.after") toolCount++;
	}

	const tsValues: number[] = [];
	for (const e of events) {
		const ts = e.timestamp_wall_ms;
		if (ts == null) continue;
		const num = Number(ts);
		if (!Number.isFinite(num)) continue;
		tsValues.push(num);
	}
	let durationMs = 0;
	if (tsValues.length > 0) {
		let minTs = tsValues[0]!;
		let maxTs = tsValues[0]!;
		for (let i = 1; i < tsValues.length; i++) {
			const v = tsValues[i]!;
			if (v < minTs) minTs = v;
			if (v > maxTs) maxTs = v;
		}
		durationMs = Math.max(0, maxTs - minTs);
	}

	const filesModified = new Set<string>();
	const filesRead = new Set<string>();
	for (const e of events) {
		if (e.type !== "tool.execute.after") continue;
		const tool = String(e.tool ?? "").toLowerCase();
		const args = e.args;
		if (args == null || typeof args !== "object") continue;
		const filePath =
			(args as Record<string, unknown>).filePath ?? (args as Record<string, unknown>).path;
		if (typeof filePath !== "string" || !filePath) continue;
		if (tool === "write" || tool === "edit") filesModified.add(filePath);
		if (tool === "read") filesRead.add(filePath);
	}

	let firstPrompt: string | undefined;
	for (const e of events) {
		if (e.type !== "user_prompt") continue;
		const text = e.prompt_text;
		if (typeof text === "string" && text.trim()) {
			firstPrompt = text.trim();
			break;
		}
	}

	return {
		firstPrompt,
		promptCount,
		toolCount,
		durationMs,
		filesModified: [...filesModified].sort(),
		filesRead: [...filesRead].sort(),
	};
}

// ---------------------------------------------------------------------------
// Main flush function
// ---------------------------------------------------------------------------

export interface FlushRawEventsOptions {
	opencodeSessionId: string;
	source?: string;
	cwd?: string | null;
	project?: string | null;
	startedAt?: string | null;
	maxEvents?: number | null;
	allowEmptyFlush?: boolean;
}

/**
 * Flush raw events for a single session through the ingest pipeline.
 *
 * 1. Reads unflushed raw events from the store
 * 2. Creates/claims a flush batch for idempotency
 * 3. Builds session context and IngestPayload
 * 4. Calls ingest()
 * 5. Updates flush state on success; records failure details on error
 *
 * Port of flush_raw_events() from raw_event_flush.py.
 */
export async function flushRawEvents(
	store: MemoryStore,
	ingestOpts: IngestOptions,
	opts: FlushRawEventsOptions,
): Promise<{ flushed: number; updatedState: number }> {
	let { source = "opencode", cwd, project, startedAt } = opts;
	const { opencodeSessionId, maxEvents } = opts;
	const allowEmptyFlush = opts.allowEmptyFlush ?? false;

	source = (source ?? "").trim().toLowerCase() || "opencode";

	// Resolve session metadata for missing fields
	const meta = store.rawEventSessionMeta(opencodeSessionId, source);
	if (cwd == null) cwd = (meta.cwd as string) ?? process.cwd();
	if (project == null) project = (meta.project as string) ?? null;
	if (startedAt == null) startedAt = (meta.started_at as string) ?? null;

	// Read unflushed events
	const lastFlushed = store.rawEventFlushState(opencodeSessionId, source);
	const events = store.rawEventsSinceBySeq(opencodeSessionId, source, lastFlushed, maxEvents);
	if (events.length === 0) {
		return { flushed: 0, updatedState: 0 };
	}

	// Extract event sequence range
	const eventSeqs: number[] = [];
	for (const e of events) {
		const seq = e.event_seq;
		if (seq == null) continue;
		const num = Number(seq);
		if (Number.isFinite(num)) eventSeqs.push(num);
	}
	if (eventSeqs.length === 0) {
		return { flushed: 0, updatedState: 0 };
	}

	let startEventSeq = eventSeqs[0]!;
	let lastEventSeq = eventSeqs[0]!;
	for (let i = 1; i < eventSeqs.length; i++) {
		const v = eventSeqs[i]!;
		if (v < startEventSeq) startEventSeq = v;
		if (v > lastEventSeq) lastEventSeq = v;
	}
	if (lastEventSeq < startEventSeq) {
		return { flushed: 0, updatedState: 0 };
	}

	// Get or create flush batch (idempotency guard)
	const { batchId, status } = store.getOrCreateRawEventFlushBatch(
		opencodeSessionId,
		source,
		startEventSeq,
		lastEventSeq,
		EXTRACTOR_VERSION,
	);

	// If already completed, just advance flush state
	if (status === "completed") {
		store.updateRawEventFlushState(opencodeSessionId, lastEventSeq, source);
		return { flushed: 0, updatedState: 1 };
	}

	// Claim the batch (atomic lock)
	if (!store.claimRawEventFlushBatch(batchId)) {
		return { flushed: 0, updatedState: 0 };
	}

	// Build session context
	const sessionContext: SessionContext = buildSessionContext(events);
	sessionContext.opencodeSessionId = opencodeSessionId;
	sessionContext.source = source;
	sessionContext.streamId = opencodeSessionId;
	sessionContext.flusher = "raw_events";
	sessionContext.flushBatch = {
		batch_id: batchId,
		start_event_seq: startEventSeq,
		end_event_seq: lastEventSeq,
	};

	// Build ingest payload
	const payload: IngestPayload = {
		cwd: cwd ?? undefined,
		project: project ?? undefined,
		startedAt: startedAt ?? new Date().toISOString(),
		events,
		sessionContext,
	};

	// Run ingest pipeline
	try {
		await ingest(payload, store, ingestOpts);
	} catch (exc) {
		// Record failure details on the batch
		const err = exc instanceof Error ? exc : new Error(String(exc));
		if (
			allowEmptyFlush &&
			err.message === "observer produced no storable output for raw-event flush"
		) {
			store.updateRawEventFlushBatchStatus(batchId, "completed");
			store.updateRawEventFlushState(opencodeSessionId, lastEventSeq, source);
			return { flushed: events.length, updatedState: 1 };
		}
		const provider = ingestOpts.observer?.getStatus?.()?.provider as string | undefined;
		const message = truncateErrorMessage(summarizeFlushFailure(err, provider));
		store.recordRawEventFlushBatchFailure(batchId, {
			message,
			errorType: err instanceof ObserverAuthError ? "ObserverAuthError" : err.name,
			observerProvider: provider ?? null,
			observerModel: (ingestOpts.observer?.getStatus?.()?.model as string) ?? null,
			observerRuntime: null,
		});
		throw exc;
	}

	// Success — mark batch completed and advance flush state
	store.updateRawEventFlushBatchStatus(batchId, "completed");
	store.updateRawEventFlushState(opencodeSessionId, lastEventSeq, source);
	return { flushed: events.length, updatedState: 1 };
}
